Spawn only two Goroutines per TCP stream

mertyildiran opened the pull request:
With this PR the number of Goroutines does not scale with the number the protocol dissectors and it will only spawn two Goroutines per TCP stream.

nimrod-up9 commented:
Copied from slack.

I think that we could end up skipping some bytes.
Suppose we got a Kafka message and that the HTTP extension just quit.
Let’s say this is the state of the `tcpReader`:
```go
reader.exhaustBuffer == true
reader.buffer == {1, 2, 3, 4}
reader.data == {5, 6, 7, 8}
```
Kafka extension calls `Read`. For some reason it gets only the first two bytes `(l == 2)`. `reader.exhaustBuffer == false`.
The extension calls `Read` again. This time it reads from `reader.data` and not from `reader.buffer` and returns 4 bytes.
The Kafka extension now holds the buffer `{1, 2, 5, 6, 7, 8}`.

Proposed solution: Always read from the same buffer `reader.data`, but to have the ability to rewind it. Instead of setting `reader.exhaustBuffer = true` after an extension quits we could call `reader.Rewind` (or some other name).

I think this is a start. Do not remove bytes from reader.data.
```go
func (reader *tcpReader) Read(p []byte) (int, error) {
	var msg api.TcpReaderDataMsg

	ok := true
	for ok && len(reader.data) == 0 {
		msg, ok = <-reader.msgQueue
		if msg != nil {
			reader.data = msg.GetBytes()
			reader.captureTime = msg.GetTimestamp()
		}

		if len(reader.data) > 0 {
			reader.packetsSeen += 1
		}
	}
	if !ok || len(reader.data) == 0 {
		return 0, io.EOF
	}

	l := copy(p, reader.data[reader.position:])
	reader.position += l
	// reader.progress.Feed(l) - not sure what to do with this

	return l, nil
}

func (reader *tcpReader) Rewind() {
	reader.position = 0
}
```

- We don’t want to keep in memory `reader.data` forever (until the stream is closed), so  we could still remove bytes if `reader.isProtocolIdentified() == true`. Need to be careful here not to mess up the `for ok && len(reader.data) == 0` loop.

mertyildiran commented:
I reproduced the skipping bytes issue...

mertyildiran commented:
@nimrod-up9 I believe I have fixed the skipping bytes issue with https://github.com/up9inc/mizu/commit/b0e95e2ea74ebdd26f85e880f0a42ebeeb28bb23

nimrod-up9 commented on tap/tcp_reader.go:
```diff
@@ -23,44 +21,44 @@ type tcpReader struct {
+		reader.counterPair = reader.parent.counterPairs[i]
+		b := bufio.NewReader(reader)
+		err := extension.Dissector.Dissect(b, reader, options)
+		if err == nil || reader.isProtocolIdentified() {
```
```suggestion
		if reader.isProtocolIdentified() {
```
I get that if there is an error, we still want to process through all of the other extensions because maybe the error is specific to the extension. So we don't break if `err != nil`.
Does `Dissect` guarantee that if the protocol is not identified, then it will return `err == nil`? If not, then this will cause a premature exit from `run`. If yes, then it is sufficient to only check `isProtocolIdentified`.

nimrod-up9 commented on tap/tcp_reader.go:
```diff
@@ -23,44 +21,44 @@ type tcpReader struct {
 	isClient      bool
 	isOutgoing    bool
 	msgQueue      chan api.TcpReaderDataMsg // Channel of captured reassembled tcp payload
+	buffer        []byte
```
```suggestion
	pastData        []byte
```

mertyildiran replied:
@nimrod-up9 done with https://github.com/up9inc/mizu/commit/c176f4c7377f1b5625b6c96d02f97cbbf83a0be3

nimrod-up9 commented on tap/tcp_reader.go:
```diff
@@ -81,14 +79,23 @@ func (reader *tcpReader) sendMsgIfNotClosed(msg api.TcpReaderDataMsg) {
 		msg, ok = <-reader.msgQueue
 		if msg != nil {
-			reader.data = msg.GetBytes()
+			reader.data = append(reader.data, msg.GetBytes()...)
```
Why append to `reader.data`? Its length is `0` here.

mertyildiran replied:
@nimrod-up9 fixed with https://github.com/up9inc/mizu/commit/6eb29988b7155f75053c38bbff57344745773566

nimrod-up9 commented on tap/tcp_reader.go:
```diff
@@ -81,14 +79,23 @@ func (reader *tcpReader) sendMsgIfNotClosed(msg api.TcpReaderDataMsg) {
+	return reader.parent.protocol != nil
+}
+
 func (reader *tcpReader) Read(p []byte) (int, error) {
+	if reader.exhaustBuffer {
+		reader.data = reader.buffer
+		reader.exhaustBuffer = false
+	}
```
```suggestion
func (reader *tcpReader) rewind() {
	reader.data = reader.buffer
}

func (reader *tcpReader) Read(p []byte) (int, error) {
```

No need for `reader.exhaustBuffer`.
Call `reader.rewind` where you would set `reader.exhaustBuffer = true`.

mertyildiran replied:
@nimrod-up9 done with https://github.com/up9inc/mizu/commit/8c8eb20ad6ea7324c2fa518fc836fa0ab31f4b6d

nimrod-up9 commented on tap/tcp_reader.go:
```diff
@@ -101,6 +108,9 @@ func (reader *tcpReader) Read(p []byte) (int, error) {
 	}
 
 	l := copy(p, reader.data)
+	if !reader.isProtocolIdentified() {
+		reader.buffer = append(reader.buffer, reader.data...)
+	}
```
This could duplicate data on `reader.buffer`.
```go
reader.data = reader.buffer
reader.buffer = append(reader.buffer, reader.data...)
```
is the same as
```
reader.buffer = append(reader.buffer, reader.buffer...)
```

Only run this when getting new messages (`msg.GetBytes()`).

mertyildiran replied:
@nimrod-up9 fixed with https://github.com/up9inc/mizu/commit/ce840940b3bd1012f431233742fc0c73c3f0afa7

nimrod-up9 commented on tap/tcp_reader.go:
```diff
@@ -81,14 +79,23 @@ func (reader *tcpReader) sendMsgIfNotClosed(msg api.TcpReaderDataMsg) {
+
 func (reader *tcpReader) Read(p []byte) (int, error) {
+	if reader.exhaustBuffer {
+		reader.data = reader.buffer
```
I'm worried that copying a slice could have unexpected results when appending. Because both slices are views to the same underlaying array, appending to one could overwrite appends to the other.
Not sure if this is the case here or if go is smart enough to avoid this effect.

mertyildiran replied:
@nimrod-up9 fixed with https://github.com/up9inc/mizu/commit/11e535cd27eaff11895cba920d6d857739d7ad20

nimrod-up9 commented on tap/tcp_reader.go:
```diff
@@ -101,6 +108,9 @@ func (reader *tcpReader) Read(p []byte) (int, error) {
 	}
 
 	l := copy(p, reader.data)
+	if !reader.isProtocolIdentified() {
```
Stop hanging on to bytes when the protocol is already identified is a nice optimization.
Why not also drop `reader.buffer` when the protocol is identified?

mertyildiran replied:
@nimrod-up9 it's actually done in here:

https://github.com/up9inc/mizu/blob/b0e95e2ea74ebdd26f85e880f0a42ebeeb28bb23/tap/tcp_stream.go#L72-L73

mertyildiran commented:
Changed the approach from buffering the bytes to buffering the TCP packet's (`api.TcpReaderDataMsg`) since the timestamp `captureTime` should be rewinded as well.

Fixed an issue of the early timing out of unidentified TCP streams because of uninitialized `createdAt` field.

Now the acceptance tests are passed.

nimrod-up9 commented on tap/tcp_reader.go:
```diff
@@ -15,35 +15,34 @@ import (
+	isClient        bool
+	isOutgoing      bool
+	msgQueue        chan api.TcpReaderDataMsg // Channel of captured reassembled tcp payload
+	msgBuffer       []api.TcpReaderDataMsg
+	msgBufferMaster []api.TcpReaderDataMsg
```
What's the benefit of buffering the messages over buffering the bytes?

mertyildiran replied:
> Changed the approach from buffering the bytes to buffering the TCP packet's (`api.TcpReaderDataMsg`) since the timestamp `captureTime` should be rewinded as well.

from https://github.com/up9inc/mizu/pull/1062#issuecomment-1126286372

Also it's cleaner but probably using slightly more memory.

nimrod-up9 replied:
Got it. It does make more sense.
